### PR TITLE
Allow plugins to plug rescore implementations

### LINKED ISSUE #26208: TopDocs from first pass to be accessible by custom ScoreFunctions
Continuing off of [the other issue I submitted](https://github.com/elastic/elasticsearch/issues/26105) and the surrounding [discussion thread](https://discuss.elastic.co/t/accessing-topdocs-from-a-leafscorefunction/96336),

Our score function computes a minimum score threshold by looking at the `TopDocs` from the first pass and decides whether to rescore a doc or not based on whether its first pass score is above this threshold. Now that `TopDocs` is harder to access, I'm hoping there can be a more supported and official way of accessing `TopDocs` for this purpose in the future.

### PR DESCRIPTION
**Pull request by nik9000**
edit:

This is what I actually merged:

This allows plugins to plug rescore implementations into
Elasticsearch. While this is a fairly expert thing to do I've
done my best to point folks to the QueryRescorer as one that at
least documents the tradeoffs that it makes. I've attempted to
limit the API surface area by removing `SearchContext` from the
exposed interface, instead exposing just the IndexSearcher and
`QueryShardContext`. I also tried to make some of the class names
more consistent and do some general cleanup while I was there.

I entertained the notion of moving the `QueryRescorer` to module.
After all, it'd be a wonderful test to prove that you can plug
rescore implementation into Elasticsearch if the only built in
rescore implementation is in the module. But I decided against it
because the new module would require a client jar and it'd require
moving some more things around. I think if we really want to do
it, we should do it as a followup.

I did, on the other hand, create an "example" rescore plugin which
should both be a nice example for anyone wanting to plug in their
own rescore implementation and servers as a good integration test
to make sure that you can indeed plug one in.

Closes #26208

This is the original description:

This allows plugins to plug rescore implementations into Elasticsearch. While this is a fairly expert thing to do I've done my best to point folks to the `QueryRescorer` as one that at least documents the tradeoffs that it makes. I've attempted to limit the API surface area by removing `SearchContext` and `QueryShardContext` from the exposed interface, instead exposing just the `IndexSearcher` and methods to compile queries and load field data. I also tried to make some of the class names more consistent and do some general cleanup while I was there.

I entertained the notion of moving the `QueryRescorer` to module. After all, it'd be a wonderful test to prove that you can plug rescore implementation into Elasticsearch if the **only** built in rescore implementation is in the module. But I decided against it because the new module would require a client jar and it'd require moving some more things around. I think if we really want to do it, we should do it as a followup.

I did, on the other hand, create an "example" rescore plugin which should both be a nice example for anyone wanting to plug in their own rescore implementation and servers as a good integration test to make sure that you can indeed plug one in.

Closes #26208

### CODE REVIEW

**Comment by nik9000** [comment]
@ethteck, could you have a look at this and see if it exposes everything that you'd need? I think you'r rescore implementation is going to look much more like the `QueryRescorer` then my example rescorer, but I think this shows that idea.

File: core/src/main/java/org/elasticsearch/search/rescore/RescorerBuilder.java [hunk 1]
```diff
@@ -124,12 +121,53 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
     protected abstract void doXContent(XContentBuilder builder, Params params) throws IOException;
 
-    public abstract QueryRescoreContext build(QueryShardContext context) throws IOException;
+    /**
+     * Build the {@linkplain RescoreContext} that will be used to actually
+     * execute the rescore against a particular shard.
+     */
+    public final RescoreContext buildContext(QueryShardContext context) throws IOException {
+        RescoreContext rescoreContext = innerBuildContext(new RescoreContextSupport() {
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/search/rescore/RescorerBuilder.java]: Is this really needed ? Why not keeping the original `build(QueryShardContext context)` ? Custom QueryBuilder have the same signature for the build function and the `QueryShardContext` does not expose too many things like the `SearchContext` so it should not be a problem to expose this in a plugin.

File: core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java [hunk 2]
```diff
@@ -35,15 +36,9 @@
...
-    public static final String NAME = "query";
 
     @Override
-    public String name() {
-        return NAME;
-    }
-
-    @Override
-    public TopDocs rescore(TopDocs topDocs, SearchContext context, RescoreSearchContext rescoreContext) throws IOException {
+    public TopDocs rescore(TopDocs topDocs, IndexSearcher searcher, RescoreContext rescoreContext) throws IOException {
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: I agree that exposing the SearchContext here is too sensitive but without it the rescorer cannot access the original query for instance. Maybe replace it with the `SearchSourceBuilder` ?

File: core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java [hunk 3]
```diff
@@ -212,7 +203,7 @@ public void setScoreMode(String scoreMode) {
     }
 
     @Override
-    public void extractTerms(SearchContext context, RescoreSearchContext rescoreContext, Set<Term> termsSet) {
+    public void extractTerms(SearchContext context, RescoreContext rescoreContext, Set<Term> termsSet) {
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: Maybe replace `SearchContext` with `IndexSearcher` if we don't want to expose `SearchContext` in plugins.

**Review by jimczi — COMMENTED** [review]
This looks great @nik9000 !

I made a first pass the code and left some comments regarding the objects that we expose for the custom rescorers. I agree that we should hide the `SearchContext` but the `QueryShardContext` is useful and is already exposed for custom query builder. I think we should keep it.

Currently we allow a rescorer to be define even when the primary sort is not on `_score`. IMO this is a bug and we should fail the query if it's not the case.
Rescorers relies on the fact that we'll merge sort the top docs of every shard so custom rescorers must ensure that they modify the original score of the top docs **and** that they re-sort the final top docs according to the new score. This can be done in a follow up but we should document this behavior or maybe even add asserts that ensure that top docs are properly sorted after the rescoring execution.

> I entertained the notion of moving the QueryRescorer to module. After all, it'd be a wonderful test to prove that you can plug rescore implementation into Elasticsearch if the only built in rescore implementation is in the module. But I decided against it because the new module would require a client jar and it'd require moving some more things around. I think if we really want to do it, we should do it as a followup.

+1 to move it to a module in a follow up. I wonder if we need another complete example of custom rescorer if we do that. It's hard to think about a simple rescorer that solves something useful other than the `query_rescorer`.

File: core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: Ahk! I meant to do that here.

File: core/src/main/java/org/elasticsearch/search/rescore/RescorerBuilder.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/rescore/RescorerBuilder.java]: I'm ok with passing in `QueryShardContext` but it still looks like it exposes a lot of stuff! Cluster alias and the bitsetFilterCache and stuff. But, yeah, queries have access to all of it, so why not rescore?

File: core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: Not really too sensitive, just pretty likely to break in the future, I guess. I'd be happy to include the original query somehow (just give it the `Query`, `SearchSourceBuilder`?) if it is important though I'm not sure it is important to rescorers to know the original query.

Comment (jimczi) [inline core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java]: Well that's the problem, we don't know what's important for a custom rescorer.
`SearchContext` is mutable which is why I think it's too sensitive but the same applies to `SearchSourceBuilder` so you're probably right. We can find ways to pass more information in the `RescoreContext` anyways so +1 to keep this simplification.

**Comment by nik9000** [comment]
@jimczi could you have another look?

**Review by jimczi — APPROVED** [review]
LGTM
I think enough information is exposed to build any kind of rescorer based on _score. We need to add the restriction for search sort other than _score though. I'll do it in a follow up.
@ethteck I think you can customize your rescorer at will with this change. WDYT ?

**Comment by ethteck** [comment]
This looks great! I really appreciate all the work. I haven't tested it yet, but from looking over the code this seems to provide us with all the functionality we'll need.

**Comment by nik9000** [comment]
Merged! I'll backport it to 6.x as well.

**Pull request merged**